Modules/_threadmodule.c

```diff
@@ -80,13 +145,12 @@ ThreadHandle_dealloc(ThreadHandleObject *self)
     }
     HEAD_UNLOCK(&_PyRuntime);
 
-    if (self->joinable) {
-        int ret = PyThread_detach_thread(self->handle);
-        if (ret) {
-            PyErr_SetString(ThreadError, "Failed detaching thread");
-            PyErr_WriteUnraisable(tp);
-        }
+    if (get_thread_handle_state(self) != THREAD_HANDLE_INVALID &&
+        _PyOnceFlag_CallOnce(&self->once, (_Py_once_fn_t *)detach_thread,
```

Comment:
This doesn't need synchronization because it's in the destructor. It should be okay to do something like:

```
if (self->state == THREAD_HANDLE_RUNNING) {
    if (PyThread_detach_thread(handle->handle)) {
        ...
    }
    ...
}
```

Reply:
Heh I had this and then changed it back. I figured it was better to use one pattern/abstraction everywhere, even when it isn't strictly required, to reduce cognitive load on future readers who might not be familiar with this code.